test/test_deconvolutional_layer.h

```diff
@@ -41,164 +42,147 @@ TEST(deconvolutional, setup_nnp) {
   deconvolutional_layer l(2, 2, 3, 1, 2, padding::valid, true, 1, 1,
                           backend_t::nnpack);
 
-  EXPECT_EQ(l.parallelize(), true);          // if layer can be parallelized
-  EXPECT_EQ(l.in_channels(), 3u);            // num of input tensors
-  EXPECT_EQ(l.out_channels(), 1u);           // num of output tensors
-  EXPECT_EQ(l.in_data_size(), 4u);           // size of input tensors
-  EXPECT_EQ(l.out_data_size(), 32u);         // size of output tensors
-  EXPECT_EQ(l.in_data_shape().size(), 1u);   // number of inputs shapes
-  EXPECT_EQ(l.out_data_shape().size(), 1u);  // num of output shapes
-  EXPECT_EQ(l.weights().size(), 2u);         // the wieghts vector size
-  EXPECT_EQ(l.weights_grads().size(), 2u);   // the wieghts vector size
-  EXPECT_EQ(l.inputs().size(), 3u);          // num of input edges
-  EXPECT_EQ(l.outputs().size(), 1u);         // num of outpus edges
-  EXPECT_EQ(l.in_types().size(), 3u);        // num of input data types
-  EXPECT_EQ(l.out_types().size(), 1u);       // num of output data types
-  EXPECT_EQ(l.fan_in_size(), 9u);            // num of incoming connections
-  EXPECT_EQ(l.fan_out_size(), 18u);          // num of outgoing connections
+  EXPECT_EQ(l.parallelize(), true);           // if layer can be parallelized
+  EXPECT_EQ(l.in_channels(), 1u);             // num of input tensors
+  EXPECT_EQ(l.out_channels(), 1u);            // num of output tensors
+  EXPECT_EQ(l.in_data_size(), 4u);            // size of input tensors
+  EXPECT_EQ(l.out_data_size(), 32u);          // size of output tensors
+  EXPECT_EQ(l.in_data_shape().size(), 1u);    // number of inputs shapes
+  EXPECT_EQ(l.out_data_shape().size(), 1u);   // num of output shapes
+  EXPECT_EQ(l.inputs().size(), 1u);           // num of input edges
+  EXPECT_EQ(l.outputs().size(), 1u);          // num of outpus edges
+  EXPECT_EQ(l.in_types().size(), 1u);         // num of input data types
+  EXPECT_EQ(l.out_types().size(), 1u);        // num of output data types
+  EXPECT_EQ(l.fan_in_size(), 9u);             // num of incoming connections
+  EXPECT_EQ(l.fan_out_size(), 18u);           // num of outgoing connections
+  EXPECT_EQ(l.parameters().size(), 2u);       // num of trainable parameters
+  EXPECT_EQ(l.ith_parameter(0).size(), 18u);  // size of weight parameter
+  EXPECT_EQ(l.ith_parameter(1).size(), 2u);   // size of bias parameter
   EXPECT_STREQ(l.layer_type().c_str(), "deconv");  // string with layer type
 }
 #endif
 
 TEST(deconvolutional, fprop) {
-  network<sequential> nn;
-
   deconvolutional_layer l(2, 2, 3, 1, 2);
+  l.setup(false);
```

Comment:
I'm wondering here how we could make this lazy, since from
https://github.com/tiny-dnn/tiny-dnn/blob/master/tiny_dnn/layers/layer.h#L591

I see that all the time allocate memory for the graph